### ISSUE TITLE
Support for other storage providers?

### ISSUE THREAD

**Issue by awinograd** [samuelgozi/firebase-auth-lite#7]
I'd like to use this library in a react native app where local storage doesn't exist. Would you be open to adding a `storage` option in the constructor to allow overriding the default localStorage provider?

I'd be happy to implement if so

**Comment by samuelgozi** [comment]
I do plan to support cookies sessions and might consider IndexedDB.
In order to save as many lines of code as possible, I would prefer to make it modular instead of hard coding a specific solution but I didn't dwell on it much.

What kind of storage would you like to add?

**Comment by awinograd** [comment]
I'd like to use https://github.com/react-native-community/async-storage

It's similar to localStorage except that getItem and setItem are async. The idea was to do something like:

```
class Auth {

    constructor({ storage = localStorage }) {
      this.storage = storage;
    }

    async foo() {
        const result = this.storage.getItem;
        const finalResult = await (new Promise(result) === result ? result : new Promise(result));
   }

```

**Comment by samuelgozi** [comment]
I think a good solution will be for the user to provide a class with two methods, `get` and `set`, and then call those when needed. And for local storage, since its the default we will provide a solution in the budle that is on by default.

It'll look to the user like this:
```js
const storage = {
  async get() {
    return await localStorage.getItem(...);
  }

  set() {
    return await localStorage.setItem(...);
  }

  remove() {
    return await localStorage.removeItem(...);
  }
};

new Auth({
  projectId: '...',
  apiKey: '...',
  storage,
});
```

Something like that. That will make sure its implementation agnostic and any environment running the app can replace the default localStorage persistence method.

What do you think about this solution?

**Comment by awinograd** [comment]
Yes I think that's in line with what I was hoping for!

**Comment by samuelgozi** [comment]
I have a working version locally, but I wanna start writing some tests before pushing. I'm also working on a big feature for firestore, so the tests will take some more time.
If you wish, try the other branch.

**Comment by awinograd** [comment]
@samuelgozi i'm just migrated over to using my fork (https://github.com/samuelgozi/firebase-auth-lite/compare/master...ouihealth:master) in my app. Haven't yet put it into production, but it's running against some CI right now.

If at some point you'd like me to pitch in on incorporating some of my changes, or discuss why I made some (b/c the commits are kinda rough) just let me know.

**Comment by samuelgozi** [comment]
Ok, thank you for the help.
I decided that ill start committing changes only after I have about 90%+ test coverage.
I've already done multiple changes that broke stuff, and I don't like that.
Hopefully, I'll reach that goal in the coming week.

**Comment by samuelgozi** [comment]
A fix to this issue was merged and published under version: 0.6.1.
It's still not documented.
but all you need to do to take advantage of this is to pass an object with three methods to the `Auth` constructor:
```js
const auth = new Auth({
    ...
    storage: {
        set(){ ... }
        get(){ ... }
        remove(){ ... }
    }
})
```

All methods, just like `localStorage` should return strings.
The first argument is expected to be a key, and the second should be the value with the exception of the `remove()` which doesn't receive a second argument.

If the item doesn't exist, `null` should be returned(if in doubt, just follow the signature of `localStorage`).
